ppexpect.py

```diff
@@ -0,0 +1,90 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+# Poor Python Expect.
+# Simplified expect to run a process
+
+import errno
+import fcntl
+import os
+import select
+import subprocess
+import sys
+import time
+
+__all__ = ['Process', 'TimeoutError']
+
+def set_nonblocking(fd):
+    fl = fcntl.fcntl(fd, fcntl.F_GETFL)
+    fcntl.fcntl(fd, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+
+if sys.version_info[0] < 3:
+    class TimeoutError(OSError):
+        def __init__(self):
+            super(TimeoutError, self).__init__(errno.ETIMEDOUT, None)
+else:
+    class TimeoutError(TimeoutError):
+        pass
+
+class Process:
+    def __init__(self, cmd):
+        process = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            close_fds=True,
+            universal_newlines=True,
+            )
+        set_nonblocking(process.stdout.fileno())
+
+        self.process = process
+
+    def __del__(self):
+        self.close()
+
+    def write(self, msg):
+        stdin = self.process.stdin
+        stdin.write(msg)
+        stdin.flush()
+
+    def expect(self, msg, timeout=1.0):
+        data = b''
+        start = time.time()
+        while True:
+            if type(msg) == str:
+                found = msg in data.decode()
+            else:
+                found = msg.match(data.decode())
+            if found:
+                return found
+
+            data += self.__readTimeout(start, timeout)
+
+    def __readTimeout(self, start, timeout):
+        fd = self.process.stdout.fileno()
```

Comment:
I don't think you need to do this. Can't you just go straight into the select() and if there is data to read immediately it will fall through to the readFd()?

In particular, if the process is constantly writing, it will never hit the timeout because the call to readFd() is before the timeout check.

Reply:
Yes, it makes a difference in case of already expired timeout but for the "poor" implementation I'll remove it, it's fine.